crates/sui-core/src/checkpoints/mod.rs

```diff
@@ -1382,11 +1371,55 @@ impl CheckpointSignatureAggregator {
                 );
                 Err(())
             }
-            InsertResult::QuorumReached(cert) => Ok(cert),
+            InsertResult::QuorumReached(cert) => {
+                // It should not be possible with an honest network to have split brain if we already have
+                // strong quorum against a digest. It may be possible however with equivocation.
+                self.check_for_split_brain();
+                // It is not guaranteed that signature.authority == narwhal_cert.author, but we do verify
+                // the signature so we know that the author signed the message at some point.
+                if their_digest != self.digest {
+                    self.metrics.remote_checkpoint_forks.inc();
+                    warn!(
+                        checkpoint_seq = self.summary.sequence_number,
+                        "Validator {:?} has mismatching checkpoint digest {}, we have digest {}",
+                        author.concise(),
+                        their_digest,
+                        self.digest
+                    );
+                    return Err(());
+                }
+                Ok(cert)
+            }
             InsertResult::NotEnoughVotes {
                 bad_votes: _,
                 bad_authorities: _,
-            } => Err(()),
+            } => {
+                if self.signatures_by_digest.has_quorum_for_key_with_strength(
+                    &their_digest,
+                    false, /* STRENGTH == weak */
+                ) {
+                    self.check_for_split_brain();
+                }
+                Err(())
+            }
+        }
+    }
+
+    fn check_for_split_brain(&self) {
+        let digests_with_weak_quorum: Vec<_> =
+            self.signatures_by_digest.all_with_quorum().collect();
+        if digests_with_weak_quorum.len() > 1 {
```

Comment:
it occurred to me that we didn't actually define split brain properly - it isn't "more than one weak quorum exists". Its "strong quorum can no longer be reached", which includes conditions like every validator submitting a unique signature.

So, ignoring equivocation for now, I think the right test is to check whether the sum of the stake of the largest faction + the stake that has not yet voted < 2f+1

Reply:
Ah I see. Sounds good, will fix this